File: server/src/api/titles.rs
//! `/api/titles` — port of `Api/Titles.cs` (ADR 0033, S4 #126). Not yet
//! flipped in `docker/nginx.conf` — see `crate::titles`'s module doc.
//!
//! `sync` and `refresh_metadata` both spawn background tasks that mirror the
//! .NET orchestration (status transitions, sync log entries) exactly, and
//! the network legs are real too: chapter-sync since S5 #127
//! (`crate::chapters::sync_from_source`), and Discover re-match + MU alias
//! refresh since S6 #128 (`crate::discover::match_sources`,
//! `crate::metadata::mangaupdates::series_detail`). `patch_title`'s
//! content_type-change branch still just resets `sync_status` to `"ready"`
//! without re-matching sources — matches .NET's own already-stubbed
//! `ReMatchSourcesAsync`, which has the identical TODO.

use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::auth::Claims;
use crate::chapters;
use crate::error::{AppError, AppResult};
use crate::state::AppState;
use crate::titles;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/", get(list))
        .route("/new-releases", get(new_releases))
        .route(
            "/{id}",
            get(get_title).patch(patch_title).delete(remove_title),
        )
        .route("/{id}/sync", axum::routing::post(sync_title))
        .route("/{id}/sync-log", get(get_sync_log))
        .route(
            "/{id}/refresh-metadata",
            axum::routing::post(refresh_metadata),
        )
}

#[derive(Serialize)]
pub(crate) struct TitleDto {
    id: String,
    title: String,
    description: Option<String>,
    cover_url: Option<String>,
    status: String,
    is_local: bool,
    local_path: Option<String>,
    author: Option<String>,
    year: Option<i64>,
    tags: Option<String>,
    sync_status: String,
    content_type: String,
    is_explicit: bool,
    auto_download: Option<bool>,
    reader_mode: Option<String>,
    download_dir: Option<String>,
    created_at: String,
    updated_at: String,
    total_chapters: i64,
    downloaded_chapters: i64,
    chapters_read: i64,
    has_sync_warnings: bool,
}

impl From<titles::TitleListItem> for TitleDto {
    fn from(t: titles::TitleListItem) -> Self {
        Self {
            id: t.id,
            title: t.title,
            description: t.description,
            cover_url: t.cover_url,
            status: t.status,
            is_local: t.is_local,
            local_path: t.local_path,
            author: t.author,
            year: t.year,
            tags: t.tags,
            sync_status: t.sync_status,
            content_type: t.content_type,
            is_explicit: t.is_explicit,
            auto_download: t.auto_download,
            reader_mode: t.reader_mode,
            download_dir: t.download_dir,
            created_at: t.created_at.replacen(' ', "T", 1),
            updated_at: t.updated_at.replacen(' ', "T", 1),
            total_chapters: t.total_chapters,
            downloaded_chapters: t.downloaded_chapters,
            chapters_read: t.chapters_read,
            has_sync_warnings: t.has_sync_warnings,
        }
    }
}

#[derive(Serialize)]
struct NewReleaseDto {
    chapter_id: String,
    chapter_number: f64,
    chapter_title: Option<String>,
    chapter_created_at: String,
    downloaded: bool,
    manga_id: String,
    manga_title: String,
    cover_url: Option<String>,
}

impl From<titles::NewReleaseItem> for NewReleaseDto {
    fn from(c: titles::NewReleaseItem) -> Self {
        Self {
            chapter_id: c.chapter_id,
            chapter_number: c.chapter_number,
            chapter_title: c.chapter_title,
            chapter_created_at: c.chapter_created_at.replacen(' ', "T", 1),
            downloaded: c.downloaded,
            manga_id: c.manga_id,
            manga_title: c.manga_title,
            cover_url: c.cover_url,
        }
    }
}

// ── GET / ────────────────────────────────────────────────────────────────

#[derive(Deserialize)]
struct ListQuery {
    page: Option<i64>,
    limit: Option<i64>,
    search: Option<String>,
    sort: Option<String>,
    content_type: Option<String>,
    status: Option<String>,
}

async fn list(
    claims: Claims,
    State(state): State<AppState>,
    Query(q): Query<ListQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let page = q.page.unwrap_or(1).max(1);
    let limit = q.limit.unwrap_or(20).min(100);

    let content_types = q
        .content_type
        .as_deref()
        .map(|s| s.split(',').filter(|s| !s.is_empty()).collect::<Vec<_>>())
        .filter(|v| !v.is_empty());
    let statuses = q
        .status
        .as_deref()
        .map(|s| s.split(',').filter(|s| !s.is_empty()).collect::<Vec<_>>())
        .filter(|v| !v.is_empty());
    let params = titles::ListParams {
        search: q.search.as_deref().filter(|s| !s.is_empty()),
        content_types,
        statuses,
    };

    let order_by = match q.sort.as_deref() {
        Some("title_asc") => "t.title ASC",
        Some("title_desc") => "t.title DESC",
        Some("year") => "t.year DESC, t.title ASC",
        Some(_) => "t.created_at DESC",
        None if params.search.is_some() => "t.title ASC",
        None => "t.created_at DESC",
    };

    let total =
        titles::count_titles(&state.db, &claims.user_id, claims.allow_explicit, &params).await?;
    let items = titles::list_titles(
        &state.db,
        &claims.user_id,
        claims.allow_explicit,
        &params,
        order_by,
        limit,
        (page - 1) * limit,
    )
    .await?;
    let items: Vec<TitleDto> = items.into_iter().map(TitleDto::from).collect();

    Ok(Json(
        json!({ "items": items, "total": total, "page": page, "limit": limit }),
    ))
}

// ── GET /new-releases ────────────────────────────────────────────────────

async fn new_releases(
    claims: Claims,
    State(state): State<AppState>,
) -> AppResult<Json<Vec<NewReleaseDto>>> {
    let items = titles::new_releases(&state.db, &claims.user_id, claims.allow_explicit).await?;
    Ok(Json(items.into_iter().map(NewReleaseDto::from).collect()))
}

// ── GET /{id} ────────────────────────────────────────────────────────────

async fn get_title(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<Json<TitleDto>> {
    let t = titles::get_title(&state.db, &id, &claims.user_id, claims.allow_explicit)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(TitleDto::from(t)))
}

// ── DELETE /{id} ─────────────────────────────────────────────────────────

#[derive(Deserialize)]
struct RemoveQuery {
    delete_files: Option<bool>,
}

async fn remove_title(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
    Query(q): Query<RemoveQuery>,
) -> AppResult<StatusCode> {
    let is_admin = claims.role == "admin";
    if !titles::remove_user_title(&state.db, &claims.user_id, &id).await? {
        return Err(AppError::NotFound);
    }

    if titles::count_owners(&state.db, &id).await? == 0 {
        if q.delete_files.unwrap_or(false) && is_admin {
            let paths = titles::chapter_local_paths(&state.db, &id).await?;
            for p in paths.iter().filter(|p| !p.starts_with("http")) {
                let _ = std::fs::remove_file(p);
            }
            if let Some(cover) = titles::cover_url(&state.db, &id).await? {
                if !cover.starts_with("http") {
                    let _ = std::fs::remove_file(&cover);
                }
            }
        }
        titles::delete_title(&state.db, &id).await?;
    }

    Ok(StatusCode::NO_CONTENT)
}

// ── PATCH /{id} ──────────────────────────────────────────────────────────

const VALID_CONTENT_TYPES: [&str; 4] = ["manga", "manhwa", "manhua", "novel"];

#[derive(Deserialize)]
struct PatchBody {
    auto_download: Option<bool>,
    reader_mode: Option<String>,
    download_dir: Option<String>,
    is_explicit: Option<bool>,
    cover_url: Option<String>,
    content_type: Option<String>,
}

async fn patch_title(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(body): Json<PatchBody>,
) -> AppResult<StatusCode> {
    let is_admin = claims.role == "admin";
    if !titles::is_owned(&state.db, &claims.user_id, &id).await? {
        return Err(AppError::NotFound);
    }

    if let Some(v) = body.auto_download {
        titles::update_auto_download(&state.db, &id, v).await?;
    }

    if let Some(rm) = &body.reader_mode {
        if rm != "paged" && rm != "scroll" {
            return Err(AppError::UnprocessableEntity("invalid reader_mode".into()));
        }
        titles::upsert_reader_mode(&state.db, &claims.user_id, &id, rm).await?;
    }

    if let Some(dir) = &body.download_dir {
        titles::update_download_dir(&state.db, &id, Some(dir)).await?;
    }

    if let Some(v) = body.is_explicit {
        if !is_admin {
            return Err(AppError::Forbidden);
        }
        titles::update_is_explicit(&state.db, &id, v).await?;
    }

    if let Some(cover) = &body.cover_url {
        if !is_admin {
            return Err(AppError::Forbidden);
        }
        let v = if cover.is_empty() {
            None
        } else {
            Some(cover.as_str())
        };
        titles::update_cover_url(&state.db, &id, v).await?;
    }

    if let Some(ct) = &body.content_type {
        if !is_admin {
            return Err(AppError::Forbidden);
        }
        if !VALID_CONTENT_TYPES.contains(&ct.as_str()) {
            return Err(AppError::UnprocessableEntity("invalid content_type".into()));
        }
        let current = titles::get_content_type(&state.db, &id)
            .await?
            .ok_or(AppError::NotFound)?;
        if &current != ct {
            titles::update_content_type(&state.db, &id, ct).await?;
            // ReMatchSourcesAsync stub — see module doc.
            titles::update_sync_status(&state.db, &id, "ready").await?;
        }
    }

    Ok(StatusCode::NO_CONTENT)
}

// ── POST /{id}/sync ──────────────────────────────────────────────────────

async fn sync_title(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<StatusCode> {
    let content_type = titles::owned_content_type(&state.db, &claims.user_id, &id)
        .await?
        .ok_or(AppError::NotFound)?;
    let links = titles::title_source_links(&state.db, &id).await?;
    if links.is_empty() {
        return Err(AppError::NotFound);
    }

    titles::update_sync_status(&state.db, &id, "syncing").await?;
    titles::clear_sync_log(&state.db, &id).await?;

    let db = state.db.clone();
    let http = state.http.clone();
    let plugin_host_url = state.config.plugin_host_url.clone();
    tokio::spawn(async move {
        let mut any_error = false;
        for (source, source_id) in &links {
            titles::append_sync_log(&db, &id, &format!("Syncing from {source}…")).await;
            match chapters::sync_from_source(
                &db,
                &http,
                &plugin_host_url,
                &id,
                &content_type,
                source,
                source_id,
            )
            .await
            {
                Ok(count) => {
                    titles::append_sync_log(
                        &db,
                        &id,
                        &format!("Synced {count} chapter(s) from {source}"),
                    )
                    .await
                }
                Err(e) => {
                    any_error = true;
                    titles::append_sync_log(&db, &id, &format!("Error from {source}: {e}")).await;
                }
            }
        }
        let final_status = if any_error { "error" } else { "ready" };
        let _ = titles::update_sync_status(&db, &id, final_status).await;
        titles::append_sync_log(&db, &id, "Sync complete").await;
    });

    Ok(StatusCode::ACCEPTED)
}

// ── GET /{id}/sync-log ───────────────────────────────────────────────────

async fn get_sync_log(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<Json<Vec<titles::SyncLogEntry>>> {
    if !titles::is_owned(&state.db, &claims.user_id, &id).await? {
        return Err(AppError::NotFound);
    }
    Ok(Json(titles::list_sync_log(&state.db, &id).await?))
}

// ── POST /{id}/refresh-metadata ──────────────────────────────────────────

async fn refresh_metadata(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<StatusCode> {
    if !titles::is_owned(&state.db, &claims.user_id, &id).await? {
        return Err(AppError::NotFound);
    }

    titles::clear_sync_warnings(&state.db, &id).await?;
    titles::clear_sync_log(&state.db, &id).await?;

    // MU metadata + alias refresh — only for MU-sourced titles (S6 #128).
    if let Some(mu_id_str) = titles::get_mangaupdates_id(&state.db, &id).await? {
        if let Ok(mu_id) = mu_id_str.parse::<u64>() {
            if let Ok(Some(series)) = crate::metadata::mangaupdates::series_detail(
                &state.http,
                &state.config.mangaupdates_url,
                mu_id,
            )
            .await
            {
                if let Some(cover) = &series.cover_url {
                    titles::set_cover_url_if_absent(&state.db, &id, cover).await?;
                }
                if !series.associated_names.is_empty() {
                    titles::clear_title_aliases(&state.db, &id).await?;
                    for alias in &series.associated_names {
                        titles::insert_title_alias(&state.db, &id, alias).await?;
                    }
                }
            }
        }
    }

    titles::update_sync_status(&state.db, &id, "syncing").await?;

    let db = state.db.clone();
    let http = state.http.clone();
    let plugin_host_url = state.config.plugin_host_url.clone();
    tokio::spawn(async move {
        if let Some(t) = titles::fetch_title(&db, &id, "").await.ok().flatten() {
            crate::discover::match_sources(
                &db,
                &http,
                &plugin_host_url,
                &id,
                &t.title,
                &t.content_type,
                t.is_explicit,
            )
            .await;
        }
        let _ = titles::update_sync_status(&db, &id, "ready").await;
    });

    Ok(StatusCode::ACCEPTED)
}

File: server/Api/Titles.cs
using System.Linq.Expressions;
using System.Security.Claims;
using System.Text.Json;
using System.Text.Json.Serialization;
using ArrghServer.Data;
using ArrghServer.Data.Models;
using ArrghServer.Services;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace ArrghServer.Api;

public static class Titles
{
    public static RouteGroupBuilder MapTitlesRoutes(this RouteGroupBuilder group)
    {
        group.MapGet("/", ListTitles).RequireAuthorization().WithSummary("List library titles (paginated, searchable)");
        group.MapGet("/new-releases", NewReleases).RequireAuthorization().WithSummary("Titles with new unread chapters");
        group.MapGet("/{id}", GetTitle).RequireAuthorization().WithSummary("Get a single title with chapters");
        group.MapDelete("/{id}", RemoveTitle).RequireAuthorization().WithSummary("Remove title from library");
        group.MapPatch("/{id}", PatchTitle).RequireAuthorization().WithSummary("Update title settings (content_type, reader_mode, auto_download…)");
        group.MapPost("/{id}/sync", SyncTitle).RequireAuthorization().WithSummary("Trigger chapter sync for a title");
        group.MapGet("/{id}/sync-log", GetSyncLog).RequireAuthorization().WithSummary("Get sync log entries for a title");
        group.MapPost("/{id}/refresh-metadata", RefreshMetadata).RequireAuthorization().WithSummary("Re-fetch metadata from the title's metadata authority");
        return group;
    }

    // -------------------------------------------------------------------------

    static async Task<IResult> ListTitles(
        ClaimsPrincipal principal,
        AppDbContext db,
        [FromQuery] int? page,
        [FromQuery] int? limit,
        [FromQuery] string? search)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";
        var p = Math.Max(page ?? 1, 1);
        var l = Math.Min(limit ?? 20, 100);

        var baseQuery = db.Titles
            .Where(m => m.UserTitles.Any(ut => ut.UserId == userId)
                     && (m.IsExplicit == false || allowExplicit));

        if (!string.IsNullOrEmpty(search))
            baseQuery = baseQuery.Where(m => EF.Functions.Like(m.TitleName, $"%{search}%"));

        var total = await baseQuery.CountAsync();

        var ordered = string.IsNullOrEmpty(search)
            ? baseQuery.OrderByDescending(m => m.UpdatedAt)
            : baseQuery.OrderBy(m => m.TitleName);

        var items = await ordered
            .Skip((p - 1) * l)
            .Take(l)
            .Select(TitleProjection(userId, db))
            .ToListAsync();

        return Results.Ok(new { items, total, page = p, limit = l });
    }

    static async Task<IResult> NewReleases(ClaimsPrincipal principal, AppDbContext db)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";

        var items = await db.Chapters
            .Where(c => c.IsNew
                     && c.Title.UserTitles.Any(ut => ut.UserId == userId)
                     && (c.Title.IsExplicit == false || allowExplicit))
            .OrderByDescending(c => c.CreatedAt)
            .Take(30)
            .Select(c => new NewReleaseItem
            {
                ChapterId = c.Id,
                ChapterNumber = c.Number,
                ChapterTitle = c.ChapterTitle,
                ChapterCreatedAt = c.CreatedAt,
                Downloaded = c.Downloaded,
                MangaId = c.TitleId,
                MangaTitle = c.Title.TitleName,
                CoverUrl = c.Title.CoverUrl,
            })
            .ToListAsync();

        return Results.Ok(items);
    }

    static async Task<IResult> GetTitle(string id, ClaimsPrincipal principal, AppDbContext db)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;
        var allowExplicit = principal.FindFirstValue("allow_explicit") == "true";

        var title = await db.Titles
            .Where(m => m.Id == id
                     && m.UserTitles.Any(ut => ut.UserId == userId)
                     && (m.IsExplicit == false || allowExplicit))
            .Select(TitleProjection(userId, db))
            .FirstOrDefaultAsync();

        if (title is null) return Results.NotFound();
        return Results.Ok(title);
    }

    internal static Task<TitleListItem?> FetchTitleAsync(AppDbContext db, string titleId, string userId) =>
        db.Titles.Where(t => t.Id == titleId).Select(TitleProjection(userId, db)).FirstOrDefaultAsync();

    // Returns an expression tree so EF can translate the full projection to SQL.
    // Must be Expression<Func<...>>, NOT a plain method — a plain method is evaluated
    // client-side after loading the entity, leaving navigations empty.
    static Expression<Func<Title, TitleListItem>> TitleProjection(string userId, AppDbContext db) =>
        m => new TitleListItem
        {
            Id = m.Id,
            Title = m.TitleName,
            Description = m.Description,
            CoverUrl = m.CoverUrl,
            Status = m.Status,
            IsLocal = !m.TitleSources.Any(),
            LocalPath = m.LocalPath,
            Author = m.Author,
            Year = m.Year,
            Tags = m.Tags,
            SyncStatus = m.SyncStatus,
            ContentType = m.ContentType,
            IsExplicit = m.IsExplicit,
            AutoDownload = m.AutoDownload,
            ReaderMode = db.UserTitleSettings
                .Where(s => s.TitleId == m.Id && s.UserId == userId)
                .Select(s => s.ReaderMode)
                .FirstOrDefault(),
            DownloadDir = m.DownloadDir,
            CreatedAt = m.CreatedAt,
            UpdatedAt = m.UpdatedAt,
            TotalChapters = m.Chapters.LongCount(),
            DownloadedChapters = m.Chapters.LongCount(c => c.Downloaded),
            ChaptersRead = db.ReadProgresses.LongCount(rp =>
                rp.UserId == userId && rp.Completed && rp.Chapter.TitleId == m.Id),
            HasSyncWarnings = m.SyncWarnings.Any(),
        };

    static async Task<IResult> RemoveTitle(
        string id,
        ClaimsPrincipal principal,
        AppDbContext db,
        [FromQuery] bool? delete_files)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;
        var isAdmin = principal.FindFirstValue(ClaimTypes.Role) == "admin";

        var userTitle = await db.UserTitles.FirstOrDefaultAsync(ut => ut.UserId == userId && ut.TitleId == id);
        if (userTitle is null) return Results.NotFound();

        db.UserTitles.Remove(userTitle);
        await db.SaveChangesAsync();

        var remaining = await db.UserTitles.CountAsync(ut => ut.TitleId == id);
        if (remaining == 0)
        {
            if ((delete_files ?? false) && isAdmin)
            {
                var chapterPaths = await db.Chapters
                    .Where(c => c.TitleId == id && c.LocalPath != null)
                    .Select(c => c.LocalPath!)
                    .ToListAsync();

                var coverUrl = await db.Titles
                    .Where(t => t.Id == id)
                    .Select(t => t.CoverUrl)
                    .FirstOrDefaultAsync();

                foreach (var path in chapterPaths.Where(p => !p.StartsWith("http")))
                    try { File.Delete(path); } catch { }

                if (coverUrl is not null && !coverUrl.StartsWith("http"))
                    try { File.Delete(coverUrl); } catch { }
            }

            var title = await db.Titles.FindAsync(id);
            if (title is not null)
            {
                db.Titles.Remove(title);
                await db.SaveChangesAsync();
            }
        }

        return Results.NoContent();
    }

    static async Task<IResult> PatchTitle(string id, PatchTitleBody body, ClaimsPrincipal principal, AppDbContext db)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;
        var isAdmin = principal.FindFirstValue(ClaimTypes.Role) == "admin";

        var owns = await db.UserTitles.AnyAsync(ut => ut.UserId == userId && ut.TitleId == id);
        if (!owns) return Results.NotFound();

        var title = await db.Titles.FindAsync(id);
        if (title is null) return Results.NotFound();

        if (body.AutoDownload is not null)
            title.AutoDownload = body.AutoDownload;

        if (body.ReaderMode.HasValue)
        {
            var rm = body.ReaderMode.Value.ValueKind == JsonValueKind.Null
                ? null
                : body.ReaderMode.Value.GetString();
            if (rm is not null && rm != "paged" && rm != "scroll")
                return Results.UnprocessableEntity();

            var settings = await db.UserTitleSettings.FindAsync(userId, id);
            if (settings is null)
                db.UserTitleSettings.Add(new() { UserId = userId, TitleId = id, ReaderMode = rm });
            else
                settings.ReaderMode = rm;
        }

        if (body.DownloadDir.HasValue)
            title.DownloadDir = body.DownloadDir.Value.ValueKind == JsonValueKind.Null
                ? null
                : body.DownloadDir.Value.GetString();

        if (body.IsExplicit is not null)
        {
            if (!isAdmin) return Results.Forbid();
            title.IsExplicit = body.IsExplicit.Value;
        }

        if (body.CoverUrl is not null)
        {
            if (!isAdmin) return Results.Forbid();
            title.CoverUrl = string.IsNullOrEmpty(body.CoverUrl) ? null : body.CoverUrl;
        }

        if (body.ContentType is not null)
        {
            if (!isAdmin) return Results.Forbid();
            if (!new[] { "manga", "manhwa", "manhua", "novel" }.Contains(body.ContentType))
                return Results.UnprocessableEntity();

            if (title.ContentType != body.ContentType)
            {
                title.ContentType = body.ContentType;
                await db.SaveChangesAsync();
                _ = Task.Run(() => ReMatchSourcesAsync(db, id, body.ContentType));
                return Results.NoContent();
            }
        }

        await db.SaveChangesAsync();
        return Results.NoContent();
    }

    static async Task<IResult> SyncTitle(
        string id, ClaimsPrincipal principal,
        AppDbContext db, IHttpClientFactory httpFactory, IConfiguration config,
        IServiceScopeFactory scopeFactory)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;

        var title = await db.Titles.FirstOrDefaultAsync(t =>
            t.Id == id && t.UserTitles.Any(ut => ut.UserId == userId));
        if (title is null) return Results.NotFound();

        var sourceLinks = await db.TitleSources.Where(ts => ts.TitleId == id).ToListAsync();
        if (sourceLinks.Count == 0) return Results.NotFound();

        await db.Titles.Where(t => t.Id == id)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.SyncStatus, "syncing"));
        await db.SyncLogs.Where(l => l.TitleId == id).ExecuteDeleteAsync();

        var capturedLinks = sourceLinks.Select(s => (s.Source, s.SourceId)).ToList();
        var contentType = title.ContentType;
        var pluginHostUrl = config["PluginHostUrl"] ?? "http://plugin-host:4000";

        _ = Task.Run(async () =>
        {
            using var scope = scopeFactory.CreateScope();
            var bgDb = scope.ServiceProvider.GetRequiredService<AppDbContext>();
            await SyncTitleAsync(bgDb, httpFactory, id, contentType, capturedLinks, pluginHostUrl);
        });

        return Results.StatusCode(202);
    }

    static async Task<IResult> GetSyncLog(string id, ClaimsPrincipal principal, AppDbContext db)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;

        var owns = await db.UserTitles.AnyAsync(ut => ut.UserId == userId && ut.TitleId == id);
        if (!owns) return Results.NotFound();

        var entries = await db.SyncLogs
            .Where(l => l.TitleId == id)
            .OrderBy(l => l.CreatedAt)
            .Select(l => new { l.Id, l.Message, l.CreatedAt })
            .ToListAsync();

        return Results.Ok(entries);
    }

    static async Task<IResult> RefreshMetadata(string id, ClaimsPrincipal principal, AppDbContext db, MangaUpdatesService mu, IHttpClientFactory httpFactory, IConfiguration config, IServiceScopeFactory scopeFactory)
    {
        var userId = principal.FindFirstValue(ClaimTypes.NameIdentifier)!;

        var title = await db.Titles.FirstOrDefaultAsync(m =>
            m.Id == id && m.UserTitles.Any(ut => ut.UserId == userId));
        if (title is null) return Results.NotFound();

        // Clear stale warnings and logs before retrying
        await db.SyncWarnings.Where(w => w.TitleId == id).ExecuteDeleteAsync();
        await db.SyncLogs.Where(l => l.TitleId == id).ExecuteDeleteAsync();

        // MU metadata + alias refresh (only for MU-sourced titles)
        if (title.MangaupdatesId is not null && ulong.TryParse(title.MangaupdatesId, out var muId))
        {
            var series = await mu.SeriesDetailAsync(muId);
            if (series is not null)
            {
                if (series.CoverUrl is not null && title.CoverUrl is null)
                    title.CoverUrl = series.CoverUrl;
                await db.TitleAliases.Where(a => a.TitleId == id).ExecuteDeleteAsync();
                foreach (var alias in series.AssociatedNames)
                    db.TitleAliases.Add(new() { Id = Guid.NewGuid().ToString(), TitleId = id, Alias = alias });
            }
        }

        await db.Titles.Where(t => t.Id == id)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.SyncStatus, "syncing"));
        await db.SaveChangesAsync();

        var capturedTitleId  = id;
        var capturedTitle    = title.TitleName;
        var capturedType     = title.ContentType;
        var pluginHostUrl    = config["PluginHostUrl"] ?? "http://plugin-host:4000";
        var downloadDir      = config["DownloadDir"] ?? "./downloads";

        _ = Task.Run(async () =>
        {
            using var scope = scopeFactory.CreateScope();
            var bgDb = scope.ServiceProvider.GetRequiredService<AppDbContext>();
            await Discover.MatchSourcesAsync(bgDb, httpFactory, capturedTitleId, capturedTitle, capturedType, pluginHostUrl);
            await bgDb.Titles.Where(t => t.Id == capturedTitleId)
                .ExecuteUpdateAsync(s => s.SetProperty(t => t.SyncStatus, "ready"));
        });

        return Results.StatusCode(202);
    }

    // -------------------------------------------------------------------------
    // Background stubs — replaced when plugin system is ported (ADR 0030)

    static async Task RefreshMatchAsync(AppDbContext db, string titleId, string titleName, IList<string> aliases)
    {
        await Task.CompletedTask;
    }

    static async Task ReMatchSourcesAsync(AppDbContext db, string titleId, string newContentType)
    {
        // TODO: port from Rust titles.rs patch_title content_type branch
        await db.Titles.Where(t => t.Id == titleId)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.SyncStatus, "ready"));
    }

    static async Task SyncTitleAsync(
        AppDbContext db, IHttpClientFactory httpFactory,
        string titleId, string? contentType,
        List<(string Source, string SourceId)> links, string pluginHostUrl)
    {
        var http = httpFactory.CreateClient();
        var anyError = false;

        foreach (var (source, sourceId) in links)
        {
            await AppendSyncLogAsync(db, titleId, $"Syncing from {source}…");
            try
            {
                var count = await ChapterSync.SyncFromSourceAsync(db, http, pluginHostUrl, titleId, contentType ?? "manga", source, sourceId);
                await AppendSyncLogAsync(db, titleId, $"Synced {count} chapter(s) from {source}");
            }
            catch (Exception ex)
            {
                anyError = true;
                await AppendSyncLogAsync(db, titleId, $"Error from {source}: {ex.Message}");
            }
        }

        var finalStatus = anyError && links.Count > 0 ? "error" : "ready";
        await db.Titles.Where(t => t.Id == titleId)
            .ExecuteUpdateAsync(s => s.SetProperty(t => t.SyncStatus, finalStatus));
        await AppendSyncLogAsync(db, titleId, "Sync complete");
    }

    static async Task AppendSyncLogAsync(AppDbContext db, string titleId, string message)
    {
        try
        {
            db.SyncLogs.Add(new ArrghServer.Data.Models.SyncLog
            {
                Id = Guid.NewGuid().ToString(),
                TitleId = titleId,
                Message = message,
                CreatedAt = DateTime.UtcNow,
            });
            await db.SaveChangesAsync();
        }
        catch { }
    }
}

// -------------------------------------------------------------------------
// DTOs

public class TitleListItem
{
    public string Id { get; set; } = null!;
    public string Title { get; set; } = null!;
    public string? Description { get; set; }
    public string? CoverUrl { get; set; }
    public string Status { get; set; } = null!;
    public bool IsLocal { get; set; }
    public string? LocalPath { get; set; }
    public string? Author { get; set; }
    public int? Year { get; set; }
    public string? Tags { get; set; }
    public string SyncStatus { get; set; } = null!;
    public string ContentType { get; set; } = null!;
    public bool IsExplicit { get; set; }
    public bool? AutoDownload { get; set; }
    public string? ReaderMode { get; set; }
    public string? DownloadDir { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public long TotalChapters { get; set; }
    public long DownloadedChapters { get; set; }
    public long ChaptersRead { get; set; }
    public bool HasSyncWarnings { get; set; }
}

public class NewReleaseItem
{
    public string ChapterId { get; set; } = null!;
    public double ChapterNumber { get; set; }
    public string? ChapterTitle { get; set; }
    public DateTime ChapterCreatedAt { get; set; }
    public bool Downloaded { get; set; }
    public string MangaId { get; set; } = null!;
    public string MangaTitle { get; set; } = null!;
    public string? CoverUrl { get; set; }
}

// ReaderMode/DownloadDir: JsonElement? distinguishes absent (no-op) from JSON null (clear).
// Known limitation: System.Text.Json maps JSON null → C# null for JsonElement? (HasValue=false),
// so clearing is indistinguishable from omitting — tracked as a follow-up.
public class PatchTitleBody
{
    [JsonPropertyName("auto_download")]  public bool? AutoDownload { get; set; }
    [JsonPropertyName("reader_mode")]    public JsonElement? ReaderMode { get; set; }
    [JsonPropertyName("download_dir")]   public JsonElement? DownloadDir { get; set; }
    [JsonPropertyName("is_explicit")]    public bool? IsExplicit { get; set; }
    [JsonPropertyName("cover_url")]      public string? CoverUrl { get; set; }
    [JsonPropertyName("content_type")]   public string? ContentType { get; set; }
}

